File: sorts/quick_sort.js
/**
 * Partition function using Lomuto partition scheme
 * @param {Array} arr - The array to partition
 * @param {number} low - Starting index
 * @param {number} high - Ending index
 * @returns {number} - The partition index
 */
function partition(arr, low, high) {
    let pivot = arr[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]]; // Swap
        }
    }
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]]; // Swap pivot
    return i + 1;
}

/**
 * Quick Sort algorithm
 * @param {Array} arr - The array to sort
 * @param {number} low - The starting index (optional)
 * @param {number} high - The ending index (optional)
 * @returns {Array} - The sorted array
 */
function quickSort(arr, low = 0, high = null) {
    if (high === null) high = arr.length - 1;

    if (low < high) {
        let pi = partition(arr, low, high);

        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
    return arr;
}

// Example usage
let arr = [10, 7, 8, 9, 1, 5];
console.log("Unsorted array:", arr.slice());
console.log("Sorted array:", quickSort(arr));

File: sorts/quick_sort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

/**
 * @brief Partition function using Lomuto partition scheme
 *
 * @param arr The vector to partition
 * @param low Starting index
 * @param high Ending index
 * @return int The partition index
 */
int partition(std::vector<int> &arr, int low, int high) {
  int pivot = arr[high]; // Pivot is the last element
  int i = (low - 1);     // Index of smaller element

  for (int j = low; j < high; j++) {
    // If current element is smaller than or equal to pivot
    if (arr[j] <= pivot) {
      i++;
      std::swap(arr[i], arr[j]);
    }
  }
  std::swap(arr[i + 1], arr[high]);
  return (i + 1);
}

/**
 * @brief Quick Sort algorithm
 *
 * @param arr Vector to be sorted
 * @param low Starting index
 * @param high Ending index
 */
void quick_sort(std::vector<int> &arr, int low, int high) {
  if (low < high) {
    // pi is partitioning index, arr[pi] is now at right place
    int pi = partition(arr, low, high);

    // Separately sort elements before partition and after partition
    quick_sort(arr, low, pi - 1);
    quick_sort(arr, pi + 1, high);
  }
}

int main() {
  std::vector<int> arr = {10, 7, 8, 9, 1, 5};
  int n = arr.size();

  std::cout << "Unsorted array: ";
  for (int i : arr)
    std::cout << i << " ";
  std::cout << std::endl;

  quick_sort(arr, 0, n - 1);

  std::cout << "Sorted array: ";
  for (int i : arr)
    std::cout << i << " ";
  std::cout << std::endl;

  return 0;
}
